From 67716e268cddb449d9bc417d152dba72e88e8535 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:13 -0400
Subject: [ethernet] qlcnic: Allow vlan0 traffic

Message-id: <1408626422-3941-90-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88613
O-Subject: [PATCH RHEL 7.1 089/138] qlcnic: Allow vlan0 traffic
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Rajesh Borundia <rajesh.borundia@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id cecd59d84d8e401db7860fc5fae503c965e5c470

o Adapter allows vlan0 traffic in case of SR-IOV after setting
  QLC_SRIOV_ALLOW_VLAN0 bit even though we do not add vlan0 filters.

Signed-off-by: Rajesh Borundia <rajesh.borundia@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
index cce9e14..39dc957 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
@@ -17,6 +17,7 @@
 #define QLC_MAC_STAR_DEL 7
 #define QLC_VF_FLOOD_BIT BIT_16
 #define QLC_FLOOD_MODE  0x5
+#define QLC_SRIOV_ALLOW_VLAN0 BIT_19
 
 static int qlcnic_sriov_pf_get_vport_handle(struct qlcnic_adapter *, u8);
 
@@ -330,8 +331,11 @@ static int qlcnic_sriov_pf_cfg_vlan_filtering(struct qlcnic_adapter *adapter,
   return err;
 
  cmd.req.arg[1] = 0x4;
- if (enable)
+ if (enable) {
   cmd.req.arg[1] |= BIT_16;
+  if (qlcnic_84xx_check(adapter))
+   cmd.req.arg[1] |= QLC_SRIOV_ALLOW_VLAN0;
+ }
 
  err = qlcnic_issue_cmd(adapter, &cmd);
  if (err)
-- 
1.7.1